Infinite recursion introduced in 1.0.0

The change in #845 introduced a crash in Sourcery for me.

To reproduce:

```
$ swift --version
Apple Swift version 5.2.4 (swiftlang-1103.0.32.9 clang-1103.0.32.53)
Target: x86_64-apple-darwin19.6.0
$ git clone https://github.com/krzysztofzablocki/Sourcery.git
$ cd Sourcery
$ git checkout 6f2c00fb4af7bc3d5022dff8336da9c40c40283c
$ swift build -c release
$ cd ..
$ git clone https://github.com/realm/SwiftLint.git
$ cd SwiftLint
$ git checkout 4083f1688556fb55c3a4972f69bb95ea0405ac96
$ ../Sourcery/.build/release/sourcery --sources Source/SwiftLintFramework/Rules --templates .sourcery/MasterRuleList.stencil --output .sourcery
No config file provided or it does not exist. Using command line arguments.
Scanning sources...
	loadOrParse: 1.1053340435028076
[1]    15167 bus error  ../Sourcery/.build/release/sourcery --sources Source/SwiftLintFramework/Rules
```

This crash doesn't happen with the previous commit on master: a441c57219fd933cd63ab49b162ad98cddfbda68

I think the crash is due to a stack overflow caused by infinite recursion.

Here's the backtrace I get when running Sourcery with Xcode:

<details>

<summary>Backtrace</summary>

```
* thread #12, queue = 'com.apple.root.user-interactive-qos', stop reason = EXC_BAD_ACCESS (code=2, address=0x70000a5a8ff8)
    frame #0: 0x00007fff713e1d36 libswiftCore.dylib`std::__1::pair<(anonymous namespace)::TupleCacheEntry*, swift::MetadataResponse> swift::LockingConcurrentMap<(anonymous namespace)::TupleCacheEntry, (anonymous namespace)::TupleCacheStorage>::getOrInsert<(anonymous namespace)::TupleCacheEntry::Key, swift::MetadataRequest&, swift::TargetValueWitnessTable<swift::InProcess> const*&>((anonymous namespace)::TupleCacheEntry::Key, swift::MetadataRequest&, swift::TargetValueWitnessTable<swift::InProcess> const*&) + 38
    frame #1: 0x00007fff713e18bc libswiftCore.dylib`swift_getTupleTypeMetadata + 140
    frame #2: 0x00007fff713e227e libswiftCore.dylib`swift_getTupleTypeMetadata2 + 30
    frame #3: 0x00007fff7112a9af libswiftCore.dylib`(extension in Swift):Swift.Collection.dropFirst(Swift.Int) -> A.SubSequence + 95
  * frame #4: 0x00000001000b6fb6 sourcery`String.drop(first=1, last=1, self="(T>(lhs: A<T>, rhs: A<T>) -> Bool {\n                    return false\n                }\n            \"\"\"),\n            Example(\"\"\"\n            public extension Array)") at Extensions.swift:56:28
    frame #5: 0x00000001000b6f06 sourcery`String.dropFirstAndLast(n=1, self="(T>(lhs: A<T>, rhs: A<T>) -> Bool {\n                    return false\n                }\n            \"\"\"),\n            Example(\"\"\"\n            public extension Array)") at Extensions.swift:51:16
    frame #6: 0x00000001000b7753 sourcery`String.isValidTupleName(self="(T>(lhs: A<T>, rhs: A<T>) -> Bool {\n                    return false\n                }\n            \"\"\"),\n            Example(\"\"\"\n            public extension Array)") at Extensions.swift:75:35
    frame #7: 0x000000010027890f sourcery`static Composer.actualTypeName(typeName=0x000000010644aea0, containingType=0x00000001017a87d0, unique=331 key/value pairs, typealiases=12 key/value pairs, self=SourceryFramework.Composer) at Composer.swift:348:34
    frame #8: 0x000000010027cd4f sourcery`static Composer.resolveType(typeName=0x000000010644aea0, containingType=0x00000001017a87d0, unique=331 key/value pairs, modules=0 key/value pairs, typealiases=12 key/value pairs, self=SourceryFramework.Composer) at Composer.swift:132:35
    frame #9: 0x0000000100281ec9 sourcery`closure #1 in static Composer.resolveType(typeName=0x000000010644aea0, self=SourceryFramework.Composer, containingType=0x00000001017a87d0, unique=331 key/value pairs, modules=0 key/value pairs, typealiases=12 key/value pairs) at Composer.swift:140:25
    frame #10: 0x00000001002832e4 sourcery`closure #3 in static Composer.resolveType(parameter=0x0000000106049fa0, resolveTypeWithName=0x000000010028a2d0 sourcery`partial apply forwarder for closure #1 (SourceryRuntime.TypeName) -> Swift.Optional<SourceryRuntime.Type> in static SourceryFramework.Composer.(resolveType in _8AFDFAE43BAEC17D714368D5062BEB48)(typeName: SourceryRuntime.TypeName, containingType: Swift.Optional<SourceryRuntime.Type>, unique: Swift.Dictionary<Swift.String, SourceryRuntime.Type>, modules: Swift.Dictionary<Swift.String, Swift.Dictionary<Swift.String, SourceryRuntime.Type>>, typealiases: Swift.Dictionary<Swift.String, SourceryRuntime.Typealias>) -> Swift.Optional<SourceryRuntime.Type> at <compiler-generated>) at Composer.swift:167:34
    frame #11: 0x00000001002833a2 sourcery`thunk for @callee_guaranteed (@guaranteed MethodParameter) -> (@error @owned Error) at <compiler-generated>:0
    frame #12: 0x000000010028a494 sourcery`thunk for @callee_guaranteed (@guaranteed MethodParameter) -> (@error @owned Error)partial apply at <compiler-generated>:0
    frame #13: 0x00007fff7120464d libswiftCore.dylib`(extension in Swift):Swift.Sequence.forEach((A.Element) throws -> ()) throws -> () + 381
    frame #14: 0x000000010027df0f sourcery`static Composer.resolveType(typeName=0x000000010163fb20, containingType=0x00000001017a87d0, unique=331 key/value pairs, modules=0 key/value pairs, typealiases=12 key/value pairs, self=SourceryFramework.Composer) at Composer.swift:166:32
    frame #15: 0x0000000100281ec9 sourcery`closure #1 in static Composer.resolveType(typeName=0x000000010163fb20, self=SourceryFramework.Composer, containingType=0x00000001017a87d0, unique=331 key/value pairs, modules=0 key/value pairs, typealiases=12 key/value pairs) at Composer.swift:140:25
    frame #16: 0x00000001002832e4 sourcery`closure #3 in static Composer.resolveType(parameter=0x000000010163fd90, resolveTypeWithName=0x000000010028a2d0 sourcery`partial apply forwarder for closure #1 (SourceryRuntime.TypeName) -> Swift.Optional<SourceryRuntime.Type> in static SourceryFramework.Composer.(resolveType in _8AFDFAE43BAEC17D714368D5062BEB48)(typeName: SourceryRuntime.TypeName, containingType: Swift.Optional<SourceryRuntime.Type>, unique: Swift.Dictionary<Swift.String, SourceryRuntime.Type>, modules: Swift.Dictionary<Swift.String, Swift.Dictionary<Swift.String, SourceryRuntime.Type>>, typealiases: Swift.Dictionary<Swift.String, SourceryRuntime.Typealias>) -> Swift.Optional<SourceryRuntime.Type> at <compiler-generated>) at Composer.swift:167:34
    frame #17: 0x00000001002833a2 sourcery`thunk for @callee_guaranteed (@guaranteed MethodParameter) -> (@error @owned Error) at <compiler-generated>:0
    frame #18: 0x000000010028a494 sourcery`thunk for @callee_guaranteed (@guaranteed MethodParameter) -> (@error @owned Error)partial apply at <compiler-generated>:0
    frame #19: 0x00007fff7120464d libswiftCore.dylib`(extension in Swift):Swift.Sequence.forEach((A.Element) throws -> ()) throws -> () + 381
...
931 stack frames omitted for brevity
...
    frame #950: 0x000000010027e174 sourcery`static Composer.resolveType(typeName=0x000000010132f760, containingType=0x00000001017a87d0, unique=331 key/value pairs, modules=0 key/value pairs, typealiases=12 key/value pairs, self=SourceryFramework.Composer) at Composer.swift:172:36
    frame #951: 0x0000000100281ec9 sourcery`closure #1 in static Composer.resolveType(typeName=0x000000010132f760, self=SourceryFramework.Composer, containingType=0x00000001017a87d0, unique=331 key/value pairs, modules=0 key/value pairs, typealiases=12 key/value pairs) at Composer.swift:140:25
    frame #952: 0x0000000100283854 sourcery`closure #4 in static Composer.resolveType(typeParameter=0x0000000101312780, resolveTypeWithName=0x000000010028a2d0 sourcery`partial apply forwarder for closure #1 (SourceryRuntime.TypeName) -> Swift.Optional<SourceryRuntime.Type> in static SourceryFramework.Composer.(resolveType in _8AFDFAE43BAEC17D714368D5062BEB48)(typeName: SourceryRuntime.TypeName, containingType: Swift.Optional<SourceryRuntime.Type>, unique: Swift.Dictionary<Swift.String, SourceryRuntime.Type>, modules: Swift.Dictionary<Swift.String, Swift.Dictionary<Swift.String, SourceryRuntime.Type>>, typealiases: Swift.Dictionary<Swift.String, SourceryRuntime.Typealias>) -> Swift.Optional<SourceryRuntime.Type> at <compiler-generated>) at Composer.swift:173:38
    frame #953: 0x0000000100283912 sourcery`thunk for @callee_guaranteed (@guaranteed GenericTypeParameter) -> (@error @owned Error) at <compiler-generated>:0
    frame #954: 0x000000010028a354 sourcery`partial apply for thunk for @callee_guaranteed (@guaranteed GenericTypeParameter) -> (@error @owned Error) at <compiler-generated>:0
    frame #955: 0x00007fff7120464d libswiftCore.dylib`(extension in Swift):Swift.Sequence.forEach((A.Element) throws -> ()) throws -> () + 381
    frame #956: 0x000000010027e174 sourcery`static Composer.resolveType(typeName=0x00000001017a8d10, containingType=0x00000001017a87d0, unique=331 key/value pairs, modules=0 key/value pairs, typealiases=12 key/value pairs, self=SourceryFramework.Composer) at Composer.swift:172:36
    frame #957: 0x000000010027c9e0 sourcery`closure #7 in static Composer.uniqueTypesAndFunctions(typeName=0x00000001017a8d10, containingType=0x00000001017a87d0, self=SourceryFramework.Composer, unique=331 key/value pairs, modules=0 key/value pairs, resolvedTypealiases=12 key/value pairs) at Composer.swift:81:25
    frame #958: 0x000000010027f6b6 sourcery`static Composer.resolveVariableTypes(variable=0x00000001017a1f30, type=0x00000001017a87d0, resolve=0x000000010027ca70 sourcery`partial apply forwarder for closure #7 (SourceryRuntime.TypeName, Swift.Optional<SourceryRuntime.Type>) -> Swift.Optional<SourceryRuntime.Type> in static SourceryFramework.Composer.uniqueTypesAndFunctions(SourceryRuntime.FileParserResult) -> (types: Swift.Array<SourceryRuntime.Type>, functions: Swift.Array<SourceryRuntime.Method>, typealiases: Swift.Array<SourceryRuntime.Typealias>) at <compiler-generated>, self=SourceryFramework.Composer) at Composer.swift:190:25
    frame #959: 0x000000010027f59d sourcery`closure #1 in closure #8 in static Composer.uniqueTypesAndFunctions($0=0x00000001017a1f30, self=SourceryFramework.Composer, type=0x00000001017a87d0, resolveType=0x000000010027ca70 sourcery`partial apply forwarder for closure #7 (SourceryRuntime.TypeName, Swift.Optional<SourceryRuntime.Type>) -> Swift.Optional<SourceryRuntime.Type> in static SourceryFramework.Composer.uniqueTypesAndFunctions(SourceryRuntime.FileParserResult) -> (types: Swift.Array<SourceryRuntime.Type>, functions: Swift.Array<SourceryRuntime.Method>, typealiases: Swift.Array<SourceryRuntime.Typealias>) at <compiler-generated>) at Composer.swift:88:17
    frame #960: 0x0000000100276392 sourcery`thunk for @callee_guaranteed (@guaranteed Variable) -> (@error @owned Error) at <compiler-generated>:0
    frame #961: 0x0000000100289a24 sourcery`partial apply for thunk for @callee_guaranteed (@guaranteed Variable) -> (@error @owned Error) at <compiler-generated>:0
    frame #962: 0x00007fff7120464d libswiftCore.dylib`(extension in Swift):Swift.Sequence.forEach((A.Element) throws -> ()) throws -> () + 381
    frame #963: 0x000000010027ece7 sourcery`closure #8 in static Composer.uniqueTypesAndFunctions(counter=38, types=331 values, self=SourceryFramework.Composer, resolveType=0x000000010027ca70 sourcery`partial apply forwarder for closure #7 (SourceryRuntime.TypeName, Swift.Optional<SourceryRuntime.Type>) -> Swift.Optional<SourceryRuntime.Type> in static SourceryFramework.Composer.uniqueTypesAndFunctions(SourceryRuntime.FileParserResult) -> (types: Swift.Array<SourceryRuntime.Type>, functions: Swift.Array<SourceryRuntime.Method>, typealiases: Swift.Array<SourceryRuntime.Typealias>) at <compiler-generated>, unique=331 key/value pairs) at Composer.swift:87:28
    frame #964: 0x00007fff71533181 libswiftDispatch.dylib`partial apply forwarder for reabstraction thunk helper from @callee_guaranteed (@unowned Swift.Int) -> () to @escaping @callee_guaranteed (@unowned Swift.Int) -> () + 17
    frame #965: 0x00007fff7152db94 libswiftDispatch.dylib`reabstraction thunk helper from @escaping @callee_guaranteed (@unowned Swift.Int) -> () to @callee_unowned @convention(block) (@unowned Swift.Int) -> () + 20
    frame #966: 0x0000000100e7f85b libdispatch.dylib`_dispatch_client_callout2 + 8
    frame #967: 0x0000000100e95f1d libdispatch.dylib`_dispatch_apply_invoke + 181
    frame #968: 0x0000000100e7f826 libdispatch.dylib`_dispatch_client_callout + 8
    frame #969: 0x0000000100e93573 libdispatch.dylib`_dispatch_root_queue_drain + 816
    frame #970: 0x0000000100e93e03 libdispatch.dylib`_dispatch_worker_thread2 + 127
    frame #971: 0x0000000100f0d31b libsystem_pthread.dylib`_pthread_wqthread + 220
    frame #972: 0x0000000100f0c49b libsystem_pthread.dylib`start_wqthread + 15
```

</details>

Got the same situation with the release version trying to generate tests for SwiftLint:

```
artem@agarmash-MBP SwiftLint % brew info sourcery
sourcery: stable 1.0.0 (bottled), HEAD
Meta-programming for Swift, stop writing boilerplate code
https://github.com/krzysztofzablocki/Sourcery
/usr/local/Cellar/sourcery/1.0.0 (7 files, 8.8MB) *
  Poured from bottle on 2020-11-10 at 15:07:37
From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/sourcery.rb
License: MIT
==> Requirements
Required: Xcode >= 10.2 ✔
==> Options
--HEAD
	Install HEAD version
==> Analytics
install: 821 (30 days), 3,217 (90 days), 14,445 (365 days)
install-on-request: 817 (30 days), 3,209 (90 days), 14,426 (365 days)
build-error: 0 (30 days)
artem@agarmash-MBP SwiftLint % swift --version
Apple Swift version 5.3 (swiftlang-1200.0.29.2 clang-1200.0.30.1)
Target: x86_64-apple-darwin19.4.0
artem@agarmash-MBP SwiftLint % make sourcery     
sourcery --sources Source/SwiftLintFramework/Rules --templates .sourcery/PrimaryRuleList.stencil --output .sourcery
No config file provided or it does not exist. Using command line arguments.
Scanning sources...
	loadOrParse: 0.2522190809249878
make: *** [Source/SwiftLintFramework/Models/PrimaryRuleList.swift] Bus error: 10
```

--- a/Sourcery/Utils/Array+Parallel.swift
+++ b/Sourcery/Utils/Array+Parallel.swift
@@ -5,6 +5,8 @@
 
 import Foundation
 
+private let disableParallelMapForEasierDebugging = false
+
 extension Array {
     func parallelFlatMap<T>(transform: (Element) throws -> [T]) throws -> [T] {
         return try parallelMap(transform).flatMap { $0 }
@@ -15,7 +17,7 @@ extension Array {
         let count = self.count
         let maxConcurrentJobs = ProcessInfo.processInfo.activeProcessorCount
 
-        guard count > 1 && maxConcurrentJobs > 1 else {
+        guard !disableParallelMapForEasierDebugging && count > 1 && maxConcurrentJobs > 1 else {
             // skip GCD overhead if we'd only run one at a time anyway
             return try map(transform)
         }

--- a/SourceryFramework/Sources/Parsing/FileParser.swift
+++ b/SourceryFramework/Sources/Parsing/FileParser.swift
@@ -357,8 +357,11 @@ extension FileParser {
 // MARK: - Variables
 extension FileParser {
 
-    private func inferType(from string: String) -> String? {
-        let string = string.trimmingCharacters(in: .whitespaces)
+    private func inferType(from input: String) -> String? {
+        let string = input
+            .trimmingCharacters(in: .whitespacesAndNewlines)
+            .strippingComments()
+            .trimmingCharacters(in: .whitespacesAndNewlines)
         // probably lazy property or default value with closure,
         // we expect explicit type, as we don't know return type
         guard !(string.hasPrefix("{") && string.hasSuffix(")")) else { return nil }
@@ -400,7 +403,9 @@ extension FileParser {
         } else if string.first == "[", string.last == "]" {
             //collection
             let string = string.dropFirstAndLast()
-            let items = string.commaSeparated()
+            let items = string.commaSeparated().map {
+                $0.trimmingCharacters(in: .whitespacesAndNewlines).strippingComments().trimmingCharacters(in: .whitespacesAndNewlines)
+            }
 
             func genericType(from itemsTypes: [String]) -> String {
                 let genericType: String
@@ -420,7 +425,9 @@ extension FileParser {
             if items[0].colonSeparated().count == 1 {
                 var itemsTypes = [String]()
                 for item in items {
-                    guard let type = inferType(from: item) else { return nil }
+                    guard let type = inferType(from: item) else {
+                        return nil
+                    }
                     itemsTypes.append(type)
                 }
                 return "[\(genericType(from: itemsTypes))]"
@@ -432,7 +439,9 @@ extension FileParser {
                     guard keyAndValue.count == 2,
                         let keyType = inferType(from: keyAndValue[0]),
                         let valueType = inferType(from: keyAndValue[1])
-                        else { return nil }
+                        else {
+                            return nil
+                        }
 
                     keysTypes.append(keyType)
                     valuesTypes.append(valueType)
@@ -479,7 +488,6 @@ extension FileParser {
             guard substring.hasPrefix("=") else { return nil }
 
             var substring = substring.dropFirst().trimmingCharacters(in: .whitespaces)
-            substring = substring.components(separatedBy: .newlines)[0]
 
             if substring.hasSuffix("{") {
                 substring = String(substring.dropLast()).trimmingCharacters(in: .whitespaces)
@@ -494,7 +502,8 @@ extension FileParser {
         } else {
             let declaration = extract(.key, from: source)
             // swiftlint:disable:next force_unwrapping
-            typeName = TypeName("<<unknown type, please add type attribution to variable\(declaration != nil ? " '\(declaration!)'" : "")>>")
+            Log.warning("<<unknown type, please add type attribution to variable\(declaration != nil ? " '\(declaration!)'" : "")>>")
+            typeName = TypeName("UnknownTypeSoAddTypeAttributionToVariable")
         }
 
         let setterAccessibility = self.setterAccessibility(source: source)

